crates/rethnet_rpc_server/src/lib.rs

```diff
@@ -0,0 +1,455 @@
+use std::net::{SocketAddr, TcpListener};
+use std::sync::Arc;
+
+use axum::{
+    extract::{Json, State},
+    http::StatusCode,
+    Router,
+};
+use hashbrown::HashMap;
+use rethnet_eth::remote::ZeroXPrefixedBytes;
+use tokio::sync::RwLock;
+
+use rethnet_eth::{
+    remote::{
+        client::Request as RpcRequest,
+        jsonrpc,
+        jsonrpc::{Response, ResponseData},
+        methods::MethodInvocation as EthMethodInvocation,
+        BlockSpec,
+    },
+    Address, B256, U256,
+};
+use rethnet_evm::{
+    state::{AccountModifierFn, ForkState, HybridState, StateError, SyncState},
+    AccountInfo, Bytecode, KECCAK_EMPTY,
+};
+
+mod hardhat_methods;
+pub use hardhat_methods::{reset::RpcHardhatNetworkConfig, HardhatMethodInvocation};
+
+/// an RPC method with its parameters
+#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
+#[serde(untagged)]
+#[allow(clippy::large_enum_variant)]
+pub enum MethodInvocation {
+    /// an eth_* method invocation
+    Eth(EthMethodInvocation),
+    /// a hardhat_* method invocation
+    Hardhat(HardhatMethodInvocation),
+}
+
+type StateType = Arc<RwLock<Box<dyn SyncState<StateError>>>>;
+
+fn response<T>(id: jsonrpc::Id, data: ResponseData<T>) -> (StatusCode, Json<serde_json::Value>)
+where
+    T: serde::Serialize,
+{
+    let response: Response<T> = Response {
+        jsonrpc: jsonrpc::Version::V2_0,
+        id: id.clone(),
+        data,
+    };
+    if let Ok(response) = serde_json::to_value(response) {
+        (StatusCode::OK, Json(response))
+    } else {
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            Json(
+                serde_json::to_value(Response {
+                    jsonrpc: jsonrpc::Version::V2_0,
+                    id,
+                    data: ResponseData::<T>::new_error(0, "serde_json::to_value() failed", None),
+                })
+                .expect("shouldn't happen"),
+            ),
+        )
+    }
+}
+
+/// returns the state root in effect BEFORE setting the block context, so that the caller can
+/// restore the context to that state root.
+async fn set_block_context<T>(
+    state: &StateType,
+    block_spec: BlockSpec,
+) -> Result<B256, ResponseData<T>> {
+    match (*state).read().await.state_root() {
```

Comment:
I don't think you need to dereference state
```suggestion
    match state.read().await.state_root() {
```

Reply:
Thanks for suggesting this. I did it your way in my recent changes to block spec handling [here](https://github.com/NomicFoundation/hardhat/pull/3998/commits/e3a009f9b8d68719e849b423c009f01ddcf9a4c9), and i revised the rest of the instances in https://github.com/NomicFoundation/hardhat/pull/3998/commits/e6ee53f1af9d0e089bd75454f44111d4c86143d0